libzkbob-rs/src/merkle.rs

```diff
@@ -90,18 +92,53 @@ impl<D: KeyValueDB, P: PoolParams> MerkleTree<D, P> {
         
                     if height == 0 && index >= cur_next_index {
                         cur_next_index = Self::calc_next_index(index);
+                    } else if height == constants::OUTPLUSONELOG as u32 &&
+                            (index << constants::OUTPLUSONELOG) >= cur_next_index
+                    {
+                        // in case of we hold just a commitment instead leafs
+                        cur_next_index = (index + 1) << constants::OUTPLUSONELOG;
                     }
                 }
                 cur_next_index
             }
         };
 
+        // we store first index for the partial Merkle tree support (when it synced not from start)
+        let db_first_index = db.get(DbCols::NextIndex as u32, FIRST_INDEX_KEY);
+        let first_index = match db_first_index {
+            Ok(Some(first_index)) => first_index
+                .as_slice()
+                .read_u64::<BigEndian>()
+                .ok(),
+            _ => {
+                let mut cur_first_index = u64::MAX;
+                for (k, _v) in db.iter(0) {
```

Comment:
> Isn't it the same iterator as above?
> We could checks all of the ifs there and get both first and last indices in one go

Yes, the iterators are same. But it's just a workaround and it shouldn't invoked normally because `db_next_index` and `db_first_index` should be stored and updated in the database. If one of them is absent - the iterator will invoked and needed index will recover from the actual tree nodes.
For example, the actual user state doesn't contain `db_first_index`. After the database update that index should be force calculated **_once_**(!). But here is no reason to recalculate both indexes every time the library loading.